### ISSUE TITLE
Stein Variational Gradient Descent

### ISSUE THREAD

**Issue by nilsleh** [lightning-uq-box/lightning-uq-box#30]
This issue is meant to discuss a possible implementation of Stein Variational Gradient Descent (SVGD) from ([Liu and Wang, 2016](https://arxiv.org/abs/1608.04471)). 

I recently came across this method in several papers and also remembered that it is implemented in numpyro (BNN tutorial [here](https://num.pyro.ai/en/stable/examples/stein_bnn.html)).

The main paper that caught my eye where I have seen SVGD in Earth Observation data tasks is:
- [A Bayesian Deep Learning Approach to Near-Term Climate Forecasting](https://agupubs.onlinelibrary.wiley.com/doi/full/10.1029/2022MS003058)
- Code versions [here](https://github.com/Xihaier/BDL_Climate_Emulator) and [here](https://zenodo.org/record/6822275#.ZE-6i5FBzCk)

As far as I understand the main idea of SVGD non-mathematically speaking is that you have a target density function p(x) that you want to approximate with a set of "particles" initialized by some prior distribution. Then in an iterative manner, the particles are updated deterministically such that the KL divergence between the target density function p(x) and the distribution of the particles is minimized. It is an alternative to MCMC but can be applied to large datasets since it is an Variational Inference optimization problem.

The approach assumes that the target distribution is continuous, differentiable, and square-integrable, however, it does not make explicit assumptions about the type of distribution and can therefore be used to approximate multi-modal distributions, not limiting to Gaussians like many approaches do. This personally seems like the most significant motivation for why this method is interesting, however, I also don't have any experience with it (and definitely don't understand the theoretical details yet). 

Since BNNs can be trained with SVGD, I thought it would be very cool to have an implementation of it as a pl.LightningModule. Beyond that, I think it would make for an interesting method to compare against others in various settings (for example Stefan's Bimodal regression example) and our more real-world EO datasets.

Some other repos where I have found [code](https://paperswithcode.com/paper/stein-variational-gradient-descent-a-general) (all with different ways of implementing it):
- original code from paper in [theano](https://github.com/DartML/Stein-Variational-Gradient-Descent)
- [bayestorch](https://github.com/lucadellalib/bayestorch/tree/main) defines a pytorch optimizer but custom prior and posterior modules
- [pytorch code](https://github.com/ImperialCollegeLondon/GSVGD) from Grassmann paper below also includes standard SVGD

It also seems that it is an active are of research with the following two papers being examples of extending the original algorithm:
- [A stochastic Stein Variational Newton method (2022)](https://arxiv.org/abs/2204.09039)
- [Grassmann Stein Variational Gradient Descent (2022)](https://arxiv.org/abs/2202.03297)

Let me know what you think @stde, @nm19000 .

Additional resources:
- [On Stein Variational Neural Network Ensembles 2021](http://www.gatsby.ucl.ac.uk/~balaji/udl2021/accepted-papers/UDL2021-paper-067.pdf)
